### ISSUE TITLE
Illegal moves

### ISSUE THREAD

**Issue by tpoppins** [tm512/minnow#1]
Two examples. 

Dual Xeon X5670 @2.93 GHz, 12-core, HT off, 40 GB RAM
Win 7 x64 Pro SP1
Cute Chess GUI 1.1.0 latest dev build
concurrency = 8
Minnow installed as a UCI engine, with no config options

More than half of its games are aborted due to illegal moves.

```
[Event "?"]
[Site "?"]
[Date "2018.10.01"]
[Round "?"]
[White "Minnow"]
[Black "feeks 2018-01-31 LM"]
[Result "0-1"]
[ECO "E42"]
[GameDuration "00:01:23"]
[GameEndTime "2018-10-01T13:29:46.364 Eastern Daylight Time"]
[GameStartTime "2018-10-01T13:28:22.904 Eastern Daylight Time"]
[Opening "Nimzo-Indian"]
[PlyCount "30"]
[Termination "illegal move"]
[TimeControl "40/120"]
[Variation "4.e3 c5, 5.Ne2 (Rubinstein)"]

1. d4 {book} Nf6 {book} 2. c4 {book} e6 {book} 3. Nc3 {book} Bb4 {book}
4. e3 {book} c5 {book} 5. Ne2 {book} O-O {book} 6. a3 {book} Bxc3+ {book}
7. Nxc3 {+65.44/7 2.5s} cxd4 {-0.10/6 5.1s} 8. exd4 {+150.80/6 5.0s}
Ne8 {-0.15/6 3.9s} 9. Qh5 {+0.40/6 5.0s} g6 {-0.01/6 6.3s}
10. Qxh7+ {+0.75/6 5.0s} Kxh7 11. Bh6 {-7.39/6 5.0s} Kxh6 {+12.07/7 5.1s}
12. Nb5 {+30.62/6 5.0s} Kg5 {+12.10/6 3.2s} 13. d5 {+16.73/6 5.0s}
exd5 {+12.18/5 4.8s} 14. Nc7 {+18.04/8 5.0s} Qxc7 {+16.47/6 3.7s}
15. cxd5 {+22.21/6 5.0s} Qa5+ {+16.50/7 3.7s, White makes an illegal move: a3a4}
0-1

```

```
[Event "?"]
[Site "?"]
[Date "2018.10.01"]
[Round "?"]
[White "feeks 2018-01-31 LM"]
[Black "Minnow"]
[Result "1-0"]
[ECO "B07"]
[GameDuration "00:00:56"]
[GameEndTime "2018-10-01T14:18:22.510 Eastern Daylight Time"]
[GameStartTime "2018-10-01T14:17:25.582 Eastern Daylight Time"]
[Opening "Pirc defense"]
[PlyCount "21"]
[Termination "illegal move"]
[TimeControl "40/120"]

1. e4 {book} d6 {book} 2. d4 {book} Nf6 {book} 3. Nc3 {book} g6 {book}
4. Be3 {book} a6 {book} 5. a4 {book} Bg7 {book} 6. Nf3 {book} Nc6 {-3.67/6 5.0s}
7. Bd3 {+0.18/4 8.0s} Qd7 {-1.48/5 5.0s} 8. O-O {+0.22/4 6.5s}
Nb4 {-3.42/6 5.0s} 9. Re1 {+0.31/4 6.6s} d5 {-3.38/6 2.2s}
10. Ne5 {+0.50/4 5.8s} Qd8 {-3.66/6 2.8s}
11. exd5 {+0.38/4 5.2s, Black makes an illegal move: f7f6} 1-0

```

I replay such games, but the opening of the second one reproduces illegal moves every single time.

**Comment by tm512** [comment]
Hey, sorry it took years to get around to this, but minnow has been abandoned since 2015.

I decided to finally look into this issue, because I'd kinda like to get minnow playing on FICS again, and this exact bug was causing it to throw games. It appears that my implementation of the null move heuristic was to blame. In particular it seems to happen when a position is specified via a move list through UCI like `position startpos moves e2e4 [etc...]`. minnow's movegen code is more than a bit impenetrable to me after not having worked on it for so many years, but my best guess is that the null moves are corrupting the move history stack which is causing the alpha-beta search to veer off course. I'm not entirely sure though. Maybe at some point I will dive in and try to pinpoint the exact issue, but that's probably unlikely.

In the meantime, I've reverted 54a01163f08bde7fb9971dccaf3cc83e1d6d54b2 which enabled the null move heuristic by default, and for future reference added some instructions for easily reproducing the bug through minnow's CLI.

If you're still around and willing to test it out, let me know if the issue is fixed. Assuming I can get my FICS bot account unlocked, I'll be running minnow online there and will be monitoring my `icsdrone` logs to ensure that it's not throwing games, but in my old logs, it never threw a game until the null move heuristic was enabled by default, so I'm pretty confident the bug is fixed.

**Comment by tm512** [comment]
So the null move heuristic might've been part of the problem, but I noticed some illegal moves happening even with it disabled. It was quite time-consuming but I tracked the issue down and implemented a fix in commit 215952fa977088e3d03628e0865020cf5a73edbb.

Soon I'll be doing some testing with the null move heuristic reenabled to see if it was even causing issues in the first place. Considering I switched to using hash moves instead of PV moves for move ordering just a couple commits before the NMH was added, and this bug I just fixed is specifically related to moves grabbed from the hash table, I might've misplaced the blame
